src/back/spirv/layout.rs

```diff
@@ -0,0 +1,216 @@
+
+        match self.bound {
+            0 => panic!("Bound cannot be 0"),
+            _ => words.push(self.bound),
+        }
+
+        words.push(self.instruction_schema);
+        words
+    }
+}
+
+pub(crate) struct LogicalLayout {
+    pub(crate) capabilities: Vec<Word>,
+    extensions: Vec<Word>,
+    pub(crate) ext_inst_imports: Vec<Word>,
+    pub(crate) memory_model: Vec<Word>,
+    pub(crate) entry_points: Vec<Word>,
+    pub(crate) execution_modes: Vec<Word>,
+    pub(crate) debugs: Vec<Word>,
+    pub(crate) annotations: Vec<Word>,
+    pub(crate) type_declarations: Vec<Word>,
+    pub(crate) constants: Vec<Word>,
+    pub(crate) global_variables: Vec<Word>,
+    function_declarations: Vec<Word>,
+    pub(crate) function_definitions: Vec<Word>,
+}
+
+impl LogicalLayout {
+    pub(crate) fn new() -> Self {
+        LogicalLayout {
+            capabilities: vec![],
+            extensions: vec![],
+            ext_inst_imports: vec![],
+            memory_model: vec![],
+            entry_points: vec![],
+            execution_modes: vec![],
+            debugs: vec![],
+            annotations: vec![],
+            type_declarations: vec![],
+            constants: vec![],
+            global_variables: vec![],
+            function_declarations: vec![],
+            function_definitions: vec![],
+        }
+    }
+
+    pub(crate) fn in_words(&self) -> Vec<u32> {
+        let mut words: Vec<Word> = vec![];
+
+        for capability in self.capabilities.iter() {
+            words.push(*capability);
+        }
+
+        for extension in self.extensions.iter() {
+            words.push(*extension);
+        }
+
+        for ext_inst_import in self.ext_inst_imports.iter() {
+            words.push(*ext_inst_import);
+        }
+
+        for memory_model in self.memory_model.iter() {
+            words.push(*memory_model);
+        }
+
+        for entry_point in self.entry_points.iter() {
+            words.push(*entry_point);
+        }
+
+        for execution_mode in self.execution_modes.iter() {
+            words.push(*execution_mode);
+        }
+
+        for debug in self.debugs.iter() {
+            words.push(*debug);
+        }
+
+        for annotation in self.annotations.iter() {
+            words.push(*annotation);
+        }
+
+        for type_declaration in self.type_declarations.iter() {
+            words.push(*type_declaration);
+        }
+
+        for constants in self.constants.iter() {
+            words.push(*constants);
+        }
+
+        for global_variables in self.global_variables.iter() {
+            words.push(*global_variables);
+        }
+
+        for function_declaration in self.function_declarations.iter() {
+            words.push(*function_declaration);
+        }
+
+        for function_definition in self.function_definitions.iter() {
+            words.push(*function_definition);
+        }
+
+        words
+    }
+}
+
+pub(crate) struct Instruction {
+    op: u32,
```

Comment:
any reason this isn't `Op`?

Reply:
Nope no reason, will change it to `Op` and cast it later in the `to_words`.

Reply:
would this still make sense to do here?

Reply:
Yes, I will change it back to `Op`